daphne/src/messages.rs

```diff
@@ -809,6 +817,218 @@ impl Decode for HpkeCiphertext {
+                bit_length.encode(bytes);
+            }
+            VdafType::Prio3AES128Histogram { buckets } => {
+                PRIO3_AES128_HISTOGRAM.encode(bytes);
+                encode_u24_items(bytes, &(), buckets);
+            }
+            VdafType::Poplar1AES128 { bit_length } => {
+                POPLAR1_AES128.encode(bytes);
+                bit_length.encode(bytes);
+            }
+            VdafType::NotImplemented(x) => {
+                x.encode(bytes);
+            }
+        }
+    }
+}
+
+impl Decode for VdafType {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let x = u32::decode(bytes)?;
+        match x {
+            PRIO3_AES128_COUNT => Ok(Self::Prio3AES128Count),
+            PRIO3_AES128_SUM => Ok(Self::Prio3AES128Sum {
+                bit_length: u8::decode(bytes)?,
+            }),
+            PRIO3_AES128_HISTOGRAM => Ok(Self::Prio3AES128Histogram {
+                buckets: decode_u24_items(&(), bytes)?,
+            }),
+            POPLAR1_AES128 => Ok(Self::Poplar1AES128 {
+                bit_length: u16::decode(bytes)?,
+            }),
+            _ => Ok(Self::NotImplemented(x)),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Deserialize, Serialize, Debug, PartialEq, Eq)]
+pub enum DpMechanism {
+    Reserved,
+    None,
+    NotImplemented(u8),
+}
+
+impl From<DpMechanism> for u8 {
+    fn from(dp_mech: DpMechanism) -> Self {
+        match dp_mech {
+            DpMechanism::Reserved => DP_MECHANISM_RESERVED,
+            DpMechanism::None => DP_MECHANISM_NONE,
+            DpMechanism::NotImplemented(x) => x,
+        }
+    }
+}
+
+impl Encode for DpMechanism {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        u8::from(*self).encode(bytes);
+    }
+}
+
+impl Decode for DpMechanism {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        match u8::decode(bytes)? {
+            x if x == DP_MECHANISM_RESERVED => Ok(Self::Reserved),
+            x if x == DP_MECHANISM_NONE => Ok(Self::None),
+            x => Ok(Self::NotImplemented(x)),
+        }
+    }
+}
+
+// The Differential Privacy configuration of draft-wang-ppm-dap-taskprov
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct DpConfig {
+    pub mechanism: DpMechanism,
+}
+
+impl AsRef<DpConfig> for DpConfig {
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+impl Encode for DpConfig {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        self.mechanism.encode(bytes);
+    }
+}
+
+impl Decode for DpConfig {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        Ok(Self {
+            mechanism: DpMechanism::decode(bytes)?,
+        })
+    }
+}
+
+// The VDAF configuration of draft-wang-ppm-dap-taskprov
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct VdafConfig {
+    pub dp_config: DpConfig,
+    pub vdaf_type: VdafType,
+}
+
+impl AsRef<VdafConfig> for VdafConfig {
```

Comment:
Is this needed yet? Generally I like to delay implementing traits until we need them. This file has a few cases where we implement AsRef, but in each case we use it elsewhere.

Reply:
not needed, will remove until we need it.